src/audio/igo_nr/igo_nr.c

```diff
@@ -0,0 +1,543 @@
+#include <user/igo_nr.h>
+#include <user/trace.h>
+#include <sof/common.h>
+#include <sof/debug/panic.h>
+#include <sof/drivers/ipc.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/uuid.h>
+#include <sof/list.h>
+#include <sof/platform.h>
+#include <sof/string.h>
+#include <sof/audio/buffer.h>
+#include <sof/audio/component.h>
+#include <sof/audio/pipeline.h>
+#include <sof/audio/igo_nr/igo_nr_comp.h>
+#include <sof/trace/trace.h>
+#include <sof/ut.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+static const struct comp_driver comp_igo_nr;
+
+/* 696ae2bc-2877-11eb-adc1-0242ac120002 */
+DECLARE_SOF_RT_UUID("igo-nr", igo_nr_uuid,  0x696ae2bc, 0x2877, 0x11eb, 0xad, 0xc1,
+		    0x02, 0x42, 0xac, 0x12, 0x00, 0x02);
+
+DECLARE_TR_CTX(igo_nr_tr, SOF_UUID(igo_nr_uuid), LOG_LEVEL_INFO);
+
+static void igo_nr_capture_s16(struct comp_data *cd,
+			       const struct audio_stream *source,
+			       struct audio_stream *sink,
+			       int32_t src_frames,
+			       int32_t sink_frames)
+{
+	int32_t in_nch = source->channels;
+	int32_t out_nch = sink->channels;
+	int32_t i;
+	int32_t idx_in = 0;
+	int32_t idx_out = 0;
+	int16_t *x;
+	int16_t *y;
+#if CONFIG_DEBUG
+	int32_t dbg_en = cd->config.igo_params.dump_data == 1 && out_nch > 1;
+#endif
+
+	for (i = 0; i < src_frames; i++) {
+		x = audio_stream_read_frag_s16(source, idx_in);
+		cd->in[cd->in_wpt] = (*x);
+
+		cd->in_wpt++;
+		idx_in += in_nch;
+	}
+
+	for (i = 0; i < sink_frames; i++) {
+		y = audio_stream_write_frag_s32(sink, idx_out);
+		*y = (cd->out[cd->out_rpt]);
+#if CONFIG_DEBUG
+		if (dbg_en) {
+			y = audio_stream_write_frag_s32(sink, idx_out + 1);
+			*y = (cd->in[cd->out_rpt]);
+		}
+#endif
+		cd->out_rpt++;
+		idx_out += out_nch;
+	}
+}
+
+static void igo_nr_capture_s32(struct comp_data *cd,
+			       const struct audio_stream *source,
+			       struct audio_stream *sink,
+			       int32_t src_frames,
+			       int32_t sink_frames)
+{
+	int32_t in_nch = source->channels;
+	int32_t out_nch = sink->channels;
+	int32_t i;
+	int32_t idx_in = 0;
+	int32_t idx_out = 0;
+	int32_t *x;
+	int32_t *y;
+#if CONFIG_DEBUG
+	int32_t dbg_en = cd->config.igo_params.dump_data == 1 && out_nch > 1;
+#endif
+
+	for (i = 0; i < src_frames; i++) {
+		x = audio_stream_read_frag_s32(source, idx_in);
+		cd->in[cd->in_wpt] = (*x) >> 16;
+
+		cd->in_wpt++;
+		idx_in += in_nch;
+	}
+
+	for (i = 0; i < sink_frames; i++) {
+		y = audio_stream_write_frag_s32(sink, idx_out);
+		*y = (cd->out[cd->out_rpt]) << 16;
+
+#if CONFIG_DEBUG
+		if (dbg_en) {
+			y = audio_stream_write_frag_s32(sink, idx_out + 1);
+			*y = (cd->in[cd->out_rpt]) << 16;
+		}
+#endif
+		cd->out_rpt++;
+		idx_out += out_nch;
+	}
+}
+
+static inline int32_t set_capture_func(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *sourceb;
+
+	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer,
+				  sink_list);
+
+	switch (sourceb->stream.rate) {
```

Comment:
@iuliana-prodan This is the place where sample rate being filtered.

Reply:
hmm, this seems confusing from a user perspective. To disable/enable NC based on HW params. Typically what we would do is have a amixer switch, NC on/off. That way NC disable/enable does not require pipeline restart. That way your module can also have a consistent set of params that it accepts. @chiang831 what are your thoughts?

Reply:
also @lgirdwood would like your thoughts here as well

Reply:
@cujomalainey do you mean that the component should not reject the audio stream, whose parameters are not supported by the component yet, but instead bypass them from input to output?

Reply:
I'm saying we should have a amixer switch to control whether algorithm is on. If switch is on, all hw_params not supported should be rejected with EINVAL. If switch is off then we use pass through channels but we could allow more hw_params through based on our sorted functions. I am just saying the behaviour of the pipeline should not be based of the params, but on the controls as those explain better to userspace the expected experience.

Reply:
Yeah, lets not make this magically switch on/off via an indirect sample rate switch. It has to be obvious to users if it's on/off so having a kcontrol or rejecting non 16kHz rates with an error would be required.

Reply:
thanks @cujomalainey @lgirdwood. just a few questions before jumping to code rewrite. 

1. Sample rate is a HW parameter decided by Codec, but bit width (S32_LE etc.) is not, rite?
2. To reject unsupported parameter when amixer on, or to pass through when amixer off, requires that the component keep track of amixer on/off status. Is there an existing flag? I couldn't find any reference inside tdfb or src component.

Reply:
format (i.e. s32_le) is a parameter that you have to check.

you will get a callback on `.cmd` and follow the examples to parse our whether it is a get/set. You will store the state in your comp_data struct and use that to track whether you are activated.

Reply:
Should sampling rate (i.e., `arecord -r16000`) be also checked too? that's my original thought.

Seems like there is no existing amixer switch status inside the component. How about we check `dev->state`  in prior? That is, if it's `COMP_STATE_ACTIVE` then it returns EINVAL when sample rate != 16000, if it's not `COMP_STATE_ACTIVE` the it just let everything to pass through.

Reply:
The hw_params will set the rate and format and these will all be known at trigger(start) so can be easily rejected as not supported if the rate and format is incorrect (I would also print something in the log saying the rate or format is not valid).
Additionally the kcontrol could then switch ON/OFF the processing (after the params have been validated above).

Reply:
@fuyuntsuo your hw_params are passed during the `params` callback. If you choose not to do it during that callback then you should do it in `prepare` which comes after `params.` You don't need to check your state, simply checking on prepare will make sure you are in the correct configuration.

Additionally to make things easier, we can say a switch change will not take effect until the pipeline is restarted (i.e. copy does not handle any ipc config change)

Reply:
@cujomalainey okay. Here is the change I will start to implement, please help review the flow. 

1. Add a NC control as a volsw kcontrol. Keep it as an ON/OFF switch inside igo_nr component. (default=OFF)
2. Validate hw_params in `.params` callback. If switch ON, reject with `EINVAL`. If switch OFF, pass through all channels.
3. Move other algorithm parameters to UCM TLV binary, which will be passed to library via `SOF_CTRL_CMD_BINARY`.

In this way we should be able to cover all the tuning paths (i.e., sof-ctl, UCM, kcontrol) for igo_nr.

Reply:
Re 2) I would reject all invalid configs regardless if switch was ON/OFF unless @cujomalainey we have use case for the pipeline with processing being off ?

Reply:
@fuyuntsuo please use the explicit switch control, we don't need volume slider. 

@lgirdwood the only case is trying to record in 48k with processing off. But I am wondering if maybe we should use mux due to different rates.

e.g.
                  IGO <- DAI(16k)
                 /
PCM <- mux
                 \
                  DAI(48k)

Reply:
Alternative is put SRC in front of IGO on 48k DAI

Reply:
@cujomalainey thanks got you, Btw we have dynamic pipeline now so we could just unload the IGO pipe and load a passthrough pipe for 48kHz capture from same DAI endpoint ? Both pipes would have a different PCM device and would be mutually exclusive ?

Reply:
@fuyuntsuo I think this now looks correct for IGO.
@cujomalainey are you fine with using dynamic pipelines to switch between IGO and a volume passthrough pipe from the same endpoint ?

Reply:
Hmm, that might be a bit annoying with the way we setup our config but we could get around it, we are currently expecting to be able to toggle with a UCM modifier and restarting the same pipeline. But since we can guarantee mutual exclusivity we can save memory? This sounds very beneficial. Lets sync out of band.

Reply:
@lgirdwood @cujomalainey sorry for late reply. I am still trying to figuring out a way to load the mixer kcontrol to igo_nr component. All the other reference topologies seems to be `pga` type component but ours is `feature` type.  Following is the observation with current code and topology:

1. Dmesg log [here](https://github.com/thesofproject/sof/files/6381318/dmesg.txt) - error: no scontrol for widget IGONR4.0 / error: process loading failed.
2. By removing mixer part (L63-65 of igo_nr.m4), we can capture the kcontrol by `amixer -c0 | grep -i igo` on TGL DUT. But the value change doesn't propagate to SOF FW.
3. By changing igo_nr from `feature` type to `pga` type, the DUT hangs.

@marc-hb @lbetlej @mmaka1 @plbossart @ranj063 @dbaluta @paulstelian97 @iuliana-prodan @singalsu 
Please advise. I am wondering if these mixer kcontrol thing only fits `pga` type component... Thanks in advnace!